uber/automated_emails.py

```diff
@@ -1,75 +1,290 @@
+            return False
+
+        for date_filter in self.when:
+            if not date_filter():
+                return False
+
+        return True
 
     def __repr__(self):
         return '<{}: {!r}>'.format(self.__class__.__name__, self.subject)
 
-    def prev(self, x, all_sent=None):
-        if all_sent:
-            return (x.__class__.__name__, x.id, self.ident) in all_sent
-        else:
-            with Session() as session:
-                return session.query(Email).filter_by(model=x.__class__.__name__, fk_id=x.id, ident=self.ident).first()
+    def _already_sent(self, model_inst):
+        """
+        Returns true if we have a record of previously sending this email to this model
+
+        NOTE: c.PREVIOUSLY_SENT_EMAILS is a cached property and will only update at the start of each daemon run.
+        """
+        return (model_inst.__class__.__name__, model_inst.id, self.ident) in c.PREVIOUSLY_SENT_EMAILS
 
-    def should_send(self, x, all_sent=None):
+    def send_if_possible(self, model_inst, raise_errors=False):
+        """
+        If it's OK to send an email of our category to this model instance (i.e. a particular Attendee) then send it.
+        """
         try:
-            return x.email and not self.prev(x, all_sent) and self.filter(x)
+            if self._should_send(model_inst):
+                self.really_send(model_inst)
         except:
-            log.error('unexpected error', exc_info=True)
+            log.error('error sending {!r} email to {}', self.subject, model_inst.email, exc_info=True)
+            if raise_errors:
+                raise
+
+    def _should_send(self, model_inst):
+        """
+        If True, we should generate an actual email created from our email category
+        and send it to a particular model instance.
+
+        This is determined based on a few things like:
+        1) whether we have sent this exact email out yet or not (previously_sent_emails)
+        2) whether the email category has been approved (approved_idents)
+        3) whether the model instance passed in is the same type as what we want to process
+        4) do any date-based filters exist on this email category? (i.e. send 7 days before magfest)
+        5) do any other filters exist on this email category? (i.e. only if attendee.staffing == true)
+
+        Example #1 of a model instance to check:
+          self.ident: "You {attendee.name} have registered for our event!"
+          model_inst:  class Attendee: id #4532, name: "John smith"
+
+        Example #2 of a model instance to check:
+          self.ident: "Your group {group.name} owes money"
+          model_inst:  class Group: id #1251, name: "The Fighting Mongooses"
+
+        :param model_inst: The model we've been requested to use (i.e. Attendee, Group, etc)
+
+        :return: True if we should send this email to this model instance, False if not.
+        """
+
+        if not isinstance(model_inst, self.model):
```

Comment:
just to be totally clear: the version I changed that has the lambda's evaluates each one at a time, and if any evalutate to False, it immediately short circuits and doesn't evaluate the rest.  this was important because if the earlier checks (like ```getattr(model_inst, 'email', None)```) are False, then the  later checks would throw an exception assuming that we checked for basic stuff like whether it has an email attribute.